// kindergarten/lib/errors/index.js
'use strict';

exports.__esModule = true;
exports.WrongRuleDefinition = exports.RestrictedMethodError = exports.NoSandboxError = exports.NoPurposeError = exports.NoGovernessError = exports.NoExposedMethodError = exports.ArgumentError = exports.AccessDenied = undefined;

var _AccessDenied = require('./AccessDenied');

var _AccessDenied2 = _interopRequireDefault(_AccessDenied);

var _ArgumentError = require('./ArgumentError');

var _ArgumentError2 = _interopRequireDefault(_ArgumentError);

var _NoExposedMethodError = require('./NoExposedMethodError');

var _NoExposedMethodError2 = _interopRequireDefault(_NoExposedMethodError);

var _NoGovernessError = require('./NoGovernessError');

var _NoGovernessError2 = _interopRequireDefault(_NoGovernessError);

var _NoPurposeError = require('./NoPurposeError');

var _NoPurposeError2 = _interopRequireDefault(_NoPurposeError);

var _NoSandboxError = require('./NoSandboxError');

var _NoSandboxError2 = _interopRequireDefault(_NoSandboxError);

var _RestrictedMethodError = require('./RestrictedMethodError');

var _RestrictedMethodError2 = _interopRequireDefault(_RestrictedMethodError);

var _WrongRuleDefinition = require('./WrongRuleDefinition');

var _WrongRuleDefinition2 = _interopRequireDefault(_WrongRuleDefinition);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.AccessDenied = _AccessDenied2.default;
exports.ArgumentError = _ArgumentError2.default;
exports.NoExposedMethodError = _NoExposedMethodError2.default;
exports.NoGovernessError = _NoGovernessError2.default;
exports.NoPurposeError = _NoPurposeError2.default;
exports.NoSandboxError = _NoSandboxError2.default;
exports.RestrictedMethodError = _RestrictedMethodError2.default;
exports.WrongRuleDefinition = _WrongRuleDefinition2.default;
